compiler/dyno/lib/uast/chpl-syntax-printer.cpp

```diff
@@ -0,0 +1,1065 @@
+    if (implicitOpeningKeyword && (style == BlockStyle::IMPLICIT || style == uast::BlockStyle::UNNECESSARY_KEYWORD_AND_BLOCK)) {
+      os << implicitOpeningKeyword;
+    }
+    if (style != BlockStyle::IMPLICIT) {
+      interpose(os, begin, end, "\n", "{","}");
+    } else {
+      interpose(os, begin, end, "\n");
+    }
+  }
+
+
+  template<typename T>
+  void printBlockWithStyle(std::ostream& os, BlockStyle style,  T xs, const char* implicitOpeningKeyword=nullptr) {
+    printBlockWithStyle(os, style, xs.begin(), xs.end(), implicitOpeningKeyword);
+  }
+
+  template<typename T>
+  void visitLiteral(std::ostream& ss_, const T* node) {
+    ss_ << node->text().c_str();
+  }
+
+  // TODO: Nesting
+  // TODO: Semicolons
+  // TODO: Newlines
+
+
+  struct ChplSyntaxVisitor {
+    std::ostringstream ss_;
+    ChplSyntaxVisitor(std::ostringstream ss_ = std::ostringstream()) {}
+
+    void printLinkages(const Decl* node) {
+      if (node->linkage() != Decl::Linkage::DEFAULT_LINKAGE) {
+        ss_ << kindToString(node->linkage()) << " ";
+        if (node->linkageName() != nullptr) {
+          printAst(ss_, node->linkageName());
+          ss_ << " ";
+        }
+      }
+    }
+
+    void visit(const uast::ASTNode* node) {
+      ss_ << tagToString(node->tag());
+    }
+
+    void visit(const Array* node) {
+      interpose(ss_, node->children(), ", ", "[", "]");
+    }
+
+    void visit(const As* node) {
+      printAst(ss_, node->symbol());
+      ss_ << " as ";
+      printAst(ss_, node->rename());
+    }
+
+    //Attributes
+
+    void visit(const Begin* node) {
+      ss_ << "begin ";
+      if (node->withClause()) {
+        printAst(ss_, node->withClause());
+        ss_ << " ";
+      }
+      printBlockWithStyle(ss_, node->blockStyle(), node->stmts());
+    }
+
+    void visit(const Block* node) {
+      printBlockWithStyle(ss_, node->blockStyle(), node->stmts());
+    }
+
+    void visit(const BoolLiteral* node) {
+      ss_ << (node->value() ? "true" : "false");
+    }
+
+    void visit(const BracketLoop* node) {
+      ss_ << "[";
+      if (node->index()) {
+        printAst(ss_, node->index());
+        ss_ << " in ";
+      }
+      printAst(ss_, node->iterand());
+      if (node->withClause()) {
+        ss_<< " ";
+        printAst(ss_, node->withClause());
+      }
+      ss_ << "]";
+      if (node->numStmts() > 0) {
+        ss_ << " ";
+        interpose(ss_, node->stmts(), "");
+      }
+    }
+
+    void visit(const Break* node) {
+      ss_ << "break";
+      if (node->target()) {
+        ss_<< " ";
+        printAst(ss_, node->target());
+      }
+      ss_ << ";";
+    }
+
+    //builder
+    //builderResult
+
+    void visit(const BytesLiteral* node) {
+      ss_ << "b\"" << quoteStringForC(std::string(node->str().c_str())) << '"';
+    }
+
+    void visit(const Catch* node) {
+      ss_ << "catch ";
+      if (node->error()){
+        if (node->hasParensAroundError()) {
+          ss_ << "(";
```

Comment:
Couldn't we remove this outer if by just doing `if (node->hasParensAroundError()) ss << "(";` and similar for the closing paren?

Reply:
we still need the outer `if` to check that `error` isn't `nullptr`, but we can make the change adding `if`s at the open/close parens.